src/ngx_http_auth_jwt_module.c

```diff
@@ -20,543 +20,638 @@
-		{
-			ngx_str_t email_t = ngx_char_ptr_to_str_t(r->pool, (char *)email);
+		ngx_str_t *element = ngx_array_push(claims);
 
-			set_custom_header_in_headers_out(r, &emailHeaderName, &email_t);
-		}
+		*element = values[i];
 	}
 
-	jwt_free(jwt);
-	
-	return NGX_OK;
-	
-	redirect:
-		if (jwt)
+	return NGX_CONF_OK;
+}
+
+static char *merge_extract_request_claims(ngx_conf_t *cf, ngx_command_t *cmd, void *c)
+{
+	auth_jwt_conf_t *conf = c;
+	ngx_array_t *claims = conf->extract_request_claims;
+
+	if (claims == NULL)
+	{
+		claims = ngx_array_create(cf->pool, 1, sizeof(ngx_str_t));
+		conf->extract_request_claims = claims;
+	}
+
+	return merge_extract_claims(cf, claims);
+}
+
+static char *merge_extract_response_claims(ngx_conf_t *cf, ngx_command_t *cmd, void *c)
+{
+	auth_jwt_conf_t *conf = c;
+	ngx_array_t *claims = conf->extract_response_claims;
+
+	if (claims == NULL)
+	{
+		claims = ngx_array_create(cf->pool, 1, sizeof(ngx_str_t));
+		conf->extract_response_claims = claims;
+	}
+
+	return merge_extract_claims(cf, claims);
+}
+
+static ngx_int_t handle_request(ngx_http_request_t *r)
+{
+	auth_jwt_conf_t *jwtcf = ngx_http_get_module_loc_conf(r, ngx_http_auth_jwt_module);
+
+	if (!jwtcf->enabled)
+	{
+		return NGX_DECLINED;
+	}
+	else
+	{
+		// pass through options requests without token authentication
+		if (r->method == NGX_HTTP_OPTIONS)
 		{
-			jwt_free(jwt);
+			return NGX_DECLINED;
 		}
-
-		if (jwtcf->auth_jwt_redirect)
+		else
 		{
-			r->headers_out.location = ngx_list_push(&r->headers_out.headers);
+			char *jwtPtr = get_jwt(r, jwtcf->validation_type);
 
-			if (r->headers_out.location == NULL)
+			if (jwtPtr == NULL)
 			{
-				ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
+				ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "failed to find a JWT");
+				return redirect(r, jwtcf);
 			}
-
-			r->headers_out.location->hash = 1;
-			r->headers_out.location->key.len = sizeof("Location") - 1;
-			r->headers_out.location->key.data = (u_char *) "Location";
-
-			if (r->method == NGX_HTTP_GET)
+			else
 			{
-				int loginlen;
-				char * scheme;
-				ngx_str_t server;
-				ngx_str_t uri_variable_name = ngx_string("request_uri");
-				ngx_int_t uri_variable_hash;
-				ngx_http_variable_value_t * request_uri_var;
-				ngx_str_t uri;
-				ngx_str_t uri_escaped;
-				uintptr_t escaped_len;
-
-				loginlen = jwtcf->auth_jwt_loginurl.len;
-				scheme = (r->connection->ssl) ? "https" : "http";
-				server = r->headers_in.server;
-
-				// get the URI
-				uri_variable_hash = ngx_hash_key(uri_variable_name.data, uri_variable_name.len);
-				request_uri_var = ngx_http_get_variable(r, &uri_variable_name, uri_variable_hash);
-
-				// get the URI
-				if(request_uri_var && !request_uri_var->not_found && request_uri_var->valid)
+				ngx_str_t algorithm = jwtcf->algorithm;
+				int keyLength;
+				u_char *key;
+				jwt_t *jwt = NULL;
+
+				if (algorithm.len == 0 || (algorithm.len == 5 && ngx_strncmp(algorithm.data, "HS", 2) == 0))
```

Comment:
@JoshMcCullough Any reason why these if else blocks use length comparisons instead of a just `strcmp` with constants of the possible encryption algorithms?

Reply:
Yes, because the check is if the algorithm is one of a set of `HS*` options e.g. `HS256`, `HS512`, etc.